src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,347 @@
+        PARTHENON_THROW(ss);
+      }
+    }
+  }
+
+  template <typename Adder>
+  void CheckOverridable(Adder add_to_state) {
+    std::unordered_set<std::string> cache;
+    for (auto &pair : overridable_vars) {
+      auto &var = pair.first;
+      auto &count = pair.second;
+      if (provided_vars.count(var) == 0) {
+        if (count > 1) {
+          std::stringstream ss;
+          ss << "Variable " << var
+             << " registered as overridable multiple times, but never provided."
+             << " This results in undefined behaviour as to which package will provide"
+             << " it." << std::endl;
+          PARTHENON_DEBUG_WARN(ss);
+        }
+        auto &mvec = overridable_meta[var];
+        for (auto &metadata : mvec) {
+          add_to_state(var, metadata);
+        }
+      }
+    }
+  }
+
+  bool Provided(const std::string &var) { return provided_vars.count(var) > 0; }
+};
+
+bool StateDescriptor::AddSwarmValue(const std::string &value_name,
+                                    const std::string &swarm_name, const Metadata &m) {
+  if (swarmMetadataMap_.count(swarm_name) == 0) {
+    throw std::invalid_argument("Swarm " + swarm_name + " does not exist!");
+  }
+  if (swarmValueMetadataMap_[swarm_name].count(value_name) > 0) {
+    throw std::invalid_argument("Swarm value " + value_name + " already exists!");
+  }
+  swarmValueMetadataMap_[swarm_name][value_name] = m;
+
+  return true;
+}
+
+bool StateDescriptor::AddField(const std::string &field_name, const Metadata &m_in) {
+  Metadata m = m_in; // horrible hack to force const correctness
+  if (m.IsSet(Metadata::Sparse)) {
+    int sparse_id = m.GetSparseId();
+    if (sparseMetadataMap_.count(field_name) > 0) {
+      if (sparseMetadataMap_.at(field_name).count(sparse_id) > 0) {
+        return false;
+      }
+    }
+    sparseMetadataMap_[field_name][sparse_id] = m;
+  } else {
+    const std::string &assoc = m.getAssociated();
+    if (!assoc.length()) m.Associate(field_name);
+    auto miter = metadataMap_.find(field_name);
+    if (miter != metadataMap_.end()) { // this field has already been added
+      return false;
+    } else {
+      m.Associate("");
+      metadataMap_[field_name] = m;
+    }
+  }
+  return true;
+}
+
+bool StateDescriptor::FlagsPresent(std::vector<MetadataFlag> const &flags,
+                                   bool matchAny) {
+  for (auto &pair : metadataMap_) {
+    auto &metadata = pair.second;
+    if (metadata.FlagsSet(flags, matchAny)) return true;
+  }
+  for (auto &pair : sparseMetadataMap_) {
+    auto &sparsemap = pair.second;
+    for (auto &p2 : sparsemap) {
+      if (p2.second.FlagsSet(flags, matchAny)) return true;
+    }
+  }
+  return false;
+}
+
+void StateDescriptor::ValidateMetadata() {
+  MetadataLoop([&](Metadata &m) {
+    auto dependency = m.Dependency();
+    if (dependency == Metadata::None) {
+      m.Set(Metadata::Provides);
```

Comment:
So if you don't set one of the new flags, it defaults to `Provides`?  I think that seems correct.  The only reasonable alternative would be to require that a flag be provided and throw here if none is given.  Just thought I'd bring this up in case there's a good argument for one or the other.  I think I'd lean _slightly_ toward requiring that a flag be provided, just to force some explicitness and consistency in downstream apps, but it's only a slight lean.

Reply:
Yeah I went with automatically being `Provides` because that doesn't require downstream changes and because it seems to be the intuitively "right" thing. You're right, though that, enforcing explicitness has merit.

I don't mind changing this, but I'll leave it as is unless you feel strongly.